### PR TITLE
Hammer/mutex groups

### PR DESCRIPTION
**Pull request by aaronchongth**
## What's new

* requires
  * https://github.com/open-rmf/rmf_api_msgs/pull/49
  * https://github.com/open-rmf/rmf_internal_msgs/pull/67
  * https://github.com/open-rmf/rmf_ros2/pull/339
* simple table showing mutex group information
* mechanism to manually unlocking of mutex

https://github.com/open-rmf/rmf-web/assets/5383623/b7a85c96-85ee-4fd5-871b-8de80d1889f2

## Testing

* Use `office-mutex-testing` branch of `rmf_demos`, which adds the mutex supervisor and mutex groups to office world
* Dispatch a patrol to `lounge`, `tinyRobot2` will be given the task and holds onto the mutex group `right`
* Send `tinyRobot1` to `lounge` using direct dispatch
```
ros2 run rmf_demos_tasks dispatch_patrol -p lounge -n 1 -F tinyRobot -R tinyRobot1 -st 0 --use_sim_time
```
* `tinyRobot1` will not proceed to `lounge`, until the mutex held by `tinyRobot2` is released 

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/api-server/api_server/routes/fleets.py [hunk 1]
```diff
@@ -180,3 +181,39 @@ async def recommission_robot(
...
+        or mutex_groups.locked is None
+        or mutex_group not in mutex_groups.locked
+    ):
+        raise HTTPException(
+            404,
+            f"Robot {robot_name} in fleet {name} does not have mutex group {mutex_group} locked",
+        )
+
+    logger.info(
+        f"User [{user.username}] requested to manually release mutex group {mutex_group} for {robot_name} of fleet {name}"
```
Comment (koonpeng) [inline packages/api-server/api_server/routes/fleets.py:215]: I think for now we can leave out the user as #931 will auto include that in the future.

File: packages/api-server/api_server/routes/fleets.py [hunk 2]
```diff
@@ -180,3 +181,39 @@ async def recommission_robot(
...
+    if fleet_state.robots is None or robot_name not in fleet_state.robots:
+        raise HTTPException(404, f"Robot {robot_name} not found in fleet {name}")
+    mutex_groups = fleet_state.robots[robot_name].mutex_groups
+    if (
+        mutex_groups is None
+        or mutex_groups.locked is None
+        or mutex_group not in mutex_groups.locked
+    ):
+        raise HTTPException(
+            404,
```
Comment (koonpeng) [inline packages/api-server/api_server/routes/fleets.py:210]: 400 error would be more appropriate here.

File: packages/dashboard/src/components/robots/robot-mutex-group-app.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  return (
+    <TableContainer>
+      <MutexGroupTable
+        mutexGroups={Object.values(mutexGroups)}
+        onMutexGroupClick={(_ev, mutexGroup) => {
+          setOpenMutexGroupDialog(true);
+          setSelectedMutexGroup(mutexGroup);
+        }}
+      />
+      {openMutexGroupDialog && selectedMutexGroup && selectedMutexGroup.lockedBy && (
```
Comment (koonpeng) [inline packages/dashboard/src/components/robots/robot-mutex-group-app.tsx:166]: This would cause the dialog to be unmounted early before thee animation is finished, remove these conditions to keep the dialog mounted.

File: packages/dashboard/src/components/robots/robot-mutex-group-app.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+          submitting={undefined}
+          onClose={() => {
+            setOpenMutexGroupDialog(false);
+          }}
+          onSubmit={handleUnlockMutexGroup}
+        >
+          <Typography>
+            Confirm unlock mutex group [{selectedMutexGroup.name}] for [
+            {selectedMutexGroup.lockedBy}]?
+          </Typography>
```
Comment (koonpeng) [inline packages/dashboard/src/components/robots/robot-mutex-group-app.tsx:181]: After we remove the conditons above, it will probably cause a typescript error. We can check `selectedMutexGroup` here to avoid it.

File: packages/react-components/lib/robots/mutex-group-table.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      headerName: 'Waiting',
+      editable: false,
+      valueGetter: (params: GridValueGetterParams) => params.row.requestedBy?.join(', ') ?? 'n/a',
+      flex: 1,
+      filterable: true,
+    },
+  ];
+
+  return (
+    <DataGrid
```
Comment (koonpeng) [inline packages/react-components/lib/robots/mutex-group-table.tsx:70]: Can we enable `disableVirtualization` (https://mui.com/x/react-data-grid/virtualization/#disable-virtualization) to see if there will be any noticeable impact?

The reason is because the virtualization causes problem in tests after we upgrade the deps. Virtualization itself is also counter productive for anything other than huge tables. https://github.com/mui/mui-x/issues/1781

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/react-components/lib/robots/mutex-group-table.tsx (hunk 5 above)
Comment (aaronchongth) [reply packages/react-components/lib/robots/mutex-group-table.tsx:70]: https://github.com/open-rmf/rmf-web/pull/933/commits/8da155b9f1ae8f1a248eba852ba82be9d0061468

File: packages/dashboard/src/components/robots/robot-mutex-group-app.tsx (hunk 4 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/robots/robot-mutex-group-app.tsx:177]: https://github.com/open-rmf/rmf-web/pull/933/commits/8da155b9f1ae8f1a248eba852ba82be9d0061468

File: packages/dashboard/src/components/robots/robot-mutex-group-app.tsx (hunk 3 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/robots/robot-mutex-group-app.tsx]: https://github.com/open-rmf/rmf-web/pull/933/commits/8da155b9f1ae8f1a248eba852ba82be9d0061468

File: packages/api-server/api_server/routes/fleets.py (hunk 2 above)
Comment (aaronchongth) [reply packages/api-server/api_server/routes/fleets.py]: https://github.com/open-rmf/rmf-web/pull/933/commits/8da155b9f1ae8f1a248eba852ba82be9d0061468

File: packages/api-server/api_server/routes/fleets.py (hunk 1 above)
Comment (aaronchongth) [reply packages/api-server/api_server/routes/fleets.py]: https://github.com/open-rmf/rmf-web/pull/933/commits/8da155b9f1ae8f1a248eba852ba82be9d0061468

**Review by koonpeng — APPROVED** [review]
